File: 275.h-index-ii.js
/*
 * @lc app=leetcode id=275 lang=javascript
 *
 * [275] H-Index II
 */

// @lc code=start
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function (citations) {
  const N = citations.length;
  let left = 0;
  let right = N;
  while (left < right) {
    const mid = left + Math.trunc((right - left) / 2);
    if (citations[mid] < N - mid) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  return N - right;
};
// @lc code=end

File: 275.h-index-ii.py
#
# @lc app=leetcode id=275 lang=python3
#
# [275] H-Index II
#

# Time:  O(logn)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        N = len(citations)
        left, right = 0, N - 1
        while left <= right:
            mid = left + (right - left) // 2
            if citations[mid] >= N - mid:
                right = mid - 1
            else:
                left = mid + 1
        return N - left


# @lc code=end
